### DISCUSSION TITLE
Infeasibility diagnostics

### DISCUSSION THREAD

**Discussion by Bleevoe** [do-mpc/do-mpc#312] [Ideas]
**Is your feature request related to a problem? Please describe.**
When IPOPT fails to solve the problem, especially due to restoration failure, I find it useful to get information about which (scaled) residuals are violated the most to figure out what the problem is. The FAQ has a nice section on bound violation analysis, https://www.do-mpc.com/en/latest/FAQ.html#feasibility-issues, but I'd like this functionality to be generalized to analysis of all NLP contraints.

**Describe the solution you'd like**
A major part of the solution would entail human-readable labeling of the constraints. Knowing that it's nonlinear constraint number 378 that is violated does not help me, and I would also not rely on the corresponding CasADi expression being readable (although I suppose it might be...). Ideally, the information would be presented in a format similar to "ODE RHS for state my_state_var in finite element 4 and collocation point 2 has a constraint violation of 3.12e-3" (or rather, I would expect to get the data needed to easily generate the above pretty printing).

**Comment by 4flixt** [comment]
Hi @Bleevoe, 

thank you for the suggestions. Unfortunately, there are several issues that prevent an implementation of this useful feature. Most importantly, we aim to give power-users of do-mpc the flexibility to alter the nonlinear constraints (see diagram on [this page](https://www.do-mpc.com/en/latest/api/do_mpc.controller.MPC.html#mpc)).  This makes it very difficult to have a good system for labelled constraints that is still easy to use for the user. 
I will think again in the future if we can solve both both problems.
Best,
4flixt

**Reply by Bleevoe** [reply]
Thank you for the feedback. I agree that proper handling of all corner cases is complicated. In the case of user-modified constraints, the API for that should include a parameter that allows the user to also modify the constraint label, and the default behavior could be to just take the original label and make a slight modification to indicate that it has been modified, e.g. prefix it with an asterisk. This default behavior could also suffice on its own in a first implementation (i.e. letting the user set the label is just "nice to have").

Regardless, thank you for considering this, and I understand that the complexity of this feature may ultimately prevent its realization.
